docs/type-annotations.org

```diff
@@ -0,0 +1,219 @@
+* Type annotations
+
+In Elisp users are not required to provide type annotations to their
+code.  While at many places the types can be inferred there are
+places, especially in user-defined functions, where we can not infer
+the correct type.
+
+Therefore Elsa provides users with the option to annotate their
+function definitions.  The annotations are placed in the body of the
+function inside a =declare= form:
+
+#+BEGIN_SRC elisp
+(defun add-one (x)
+  (declare (elsa (int) int))
+  (1+ x))
+#+END_SRC
+
+The =declare= form starts with =elsa= followed by one or two values.  If
+two are provided, first must be a list which is the list of input
+types and the second is the return type.
+
+If only one is provided it is taken to be the return type and the
+function must have no input arguments.
+
+Here are general guidelines on how the types are constructed.
+
+** Built-in types
+
+For built-in types with test predicates, drop the =p= or =-p= suffix to
+get the type:
+
+- =stringp= → =string=
+- =integerp= → =integer= (=int= is also accepted)
+- =markerp= → =marker=
+- =hash-table-p= → =hash-table=
+
+Some additional special types:
+
+- =t= stands for =t= and is always true
+- =nil= stands for =nil= and is always false
+- =bool= is a combination of explicitly =t= or =nil=
+
+** Nullable types
+
+By default types do not accept =nil= values.  This helps preventing
+errors where you would pass =nil= to a function expecting a value.  To
+mark a type nullable you can combine it with =nil= type using the [[id:5a21a68a-4df1-4d44-a854-1d9700858a1a][or]]
+combinator: =(or string nil)= is a type which takes any string but also
+=nil=.
+
+** Most general type
+
+A type for everything is called =mixed=.  It accepts anything and is
+always nullable (that means it accepts =nil=).  This is the default type
+for when we lack type information.
+
+** Composite (higher order) types
+
+Composite or higher-order types are types which take other types as
+arguments.
+
+Composite types usually correspond to data constructors such as =cons=,
+=list=, =vector=...
+
+- =(cons a b)= where =a= is the type of =car= and =b= is the type of =cdr=.  If
+  the =car= and =cdr= can be anything write =(cons mixed mixed)= or simply
+  =cons= for short.
+- =(list a)= where =a= is the type of items in the list.  If the list can
```

Comment:
using `tuple` probably makes more sense than to reuse `list`.

Reply:
I had another thought at `tuple` and I'm not sure about it anymore. The `tuple` solution won't let you express that the first items are of some known type where the rest is unknown. Some use cases:

- the parameters to `format` (`(list string mixed...)`)
- tagged lists where the first element is a symbol indicating how to interpret what follows (`(list symbol mixed...)`)
- poor-man structures (e.g., a person could be described as a name, a birthdate and children with `(string date person...)`

Reply:
That's also a bit related to structures like plists or alists, where alists are lists of lists where the `car` is something.  These are good points.

In the future I also want to support "constant" types, so you can explicitly require specific values, for example, using your notation

    (list (or 'regex 'syntax) string...)

could mean "this is a list of strings which should be interpreted as regular expressions if the `car` is a constant `'regex`, or as syntax descriptors if `'syntax`.

The constants will probably have a constructor `const` similar to how `defcustom` does it.  Maybe we can use that syntax?  The `defcustom` type `(repeat string)` is a list of strings `("a" "b" "c" "d" ...)` where `(list string integer symbol)` is a list of three things `("foo" 1 'bar)`.

Reply:
I suggest you reuse as much as possible from `defcustom` type definitions: it's powerful, well documented and there are examples around.

Reply:
Agreed.  At some point I actually want to write a parser for the `defcustom` definitions as well so we can automatically have the types of options without adding basically the same duplicate annotation just for elsa.

Reply:
Actually this

> The tuple solution won't let you express that the first items are of some known type where the rest is unknown.

can already be done with `(cons string (list mixed))`

Reply:
true :-)

Reply:
I don't like `repeat`, it looks very weird, especially if we want to leave out the type.

A type of `length` could be `(elsa ((repeat mixed)) int)` but also `(elsa (repeat) int)` which just looks odd.

We have `sequence` but that is actually a separate type.  `array` does not fit the notion of a list.

I really like the fact that in the new proposal `list` corresponds to the constructor exactly so that `(list "a" "b" "c")` constructs a list of three strings therefore its type is `(list string string string)`.  This matches `cons` which is nice.

So I'm also changing the `vector` type to behave like this.  Which leaves me in a tough position because now I have no idea how to express a "vector of 0 or more of type".

Reply:
What happened to your idea of using `tuple`?

Reply:
We wanted to reuse the defcustom stuff so I dived into it and I realized that I really like the constructor-as-type principle.

There's also this cool pattern-matching library which uses a similar concept: https://github.com/VincentToups/shadchen-el

Reply:
So actually maybe we can lift the `list-rest` idea from there.  Or allow a `&rest` keyword inside `list` such that `(list &rest int)` is a list of integers of any length and `(list string bool &rest int)` requires a string, a bool and then a bunch of ints.

The same can work in parallel with `vector` such that `(vector string bool &rest int)` is... you get the idea.

This might be it! :D

Reply:
How about this

-----

Lists are made by chaining cons cells together and leaving =nil= in the
last =cdr= of the last =cons=.  Therefore =(cons a (cons b (cons c nil)))=
can be understood as a list of exactly three items of type =a=, =b=, and
=c=.

Elisp provides a data constructor =list= to make creating such lists
easier.  In analogy to =cons= we provide a =list= type which corresponds
to this, such that =(list "foo" 1 :bar)= has type =(list string int
keyword)=.

In general =(list a b c ...)= is a list type of known length where =a=, =b=,
=c=... are types of the items on the 0th, 1st, 2nd... places.  It's a
proper list so the last =cons='s =cdr= is a =nil=.

Sometimes we need to express lists of unknown length which contain
items of the same thing repeatedly.  For example =(3 2 5 1 5 6 7 5)= is
a list of numbers.  Because we don't know its length beforehand we
provide a special syntax to express that a type repeats forever inside
the list.

The type =(list a b c . rest-type)= is a type similar to =list= except the
last =rest-type= repeats forever.  Therefore, we expect a known number
of items of types =a=, =b=, =c= and then any number of items of =rest-type=.
If we only speficy one type it is automatically the rest type, so that
=(list . rest-type)= holds any number of items of =rest-type=.

If the list can hold any number of anything, write =(list . mixed)= or
simply =list= for short.

Reply:
I like it very much, good job. I'm not so sure about this phrasing though:

> If we only speficy one type it is automatically the rest type, so that `(list . rest-type)` holds any number of items of `rest-type`.

If the reader doesn't pay attention, she might understand that you are talking about `(list rest-type)` and not `(list . rest-type)`. May I suggest this phrasing instead?

> If we specify no type before the dot `.` as in `(list . rest-type)`, this means the list holds any number of items of `rest-type`.

Reply:
That's a good point, I wasn't so sure myself but couldn't phrase it better.

I'm also thinking about maybe introducing some other shorter notation without the dot as that is confusing and we maybe might want to use it for dotted lists (albeit those are really rare so I wouldn't care about those very much).

Something like `(list* a)` where the star denotes a repetition like in  formal grammars and stuff.  I'd probably move the generic `*` marker before the type then.